Python/bltinmodule.c

```diff
@@ -2667,15 +2734,22 @@ builtin_sum_impl(PyObject *module, PyObject *iterable, PyObject *start)
                 int overflow;
                 value = PyLong_AsLongAndOverflow(item, &overflow);
                 if (!overflow) {
-                    f_result += (double)value;
+                    cr_result.sum += (double)value;
+                    ci_result.sum += 0.0;
                     Py_DECREF(item);
                     continue;
                 }
             }
-            if (c && isfinite(c)) {
-                f_result += c;
+            if (PyFloat_Check(item)) {
+                double value = PyFloat_AS_DOUBLE(item);
+                cr_result.sum += value;
+                ci_result.sum += 0.0;
```

Comment:
Oh yes, it does.  ``-0.0 + 0.0`` is ``0.0``.  So, it mess up with signs.

Personally, I'm fan of mixed-mode rules for complex arithmetic (see https://github.com/skirpichev/cpython/pull/1), but currently CPython does type-casts for int/floats doing addition: ``complex(r,i)+float(a)==complex(r,i)+complex(a,0.0)``.  Thus, unfortunately this line seems to be required.

Reply:
I see. I don't really like floating-point arithmetic so I'm usually more a fixed-point one. Thanks for the clarification!

Reply:
Not sure if it worth, but I can add a comment.

In fact, probably it's not well documented in CPython docs.  People might expect different things (e.g. most implementations of the ``_Complex`` type in C don't do type-cast in case ``_Complex + double``).

Reply:
> Not sure if it worth, but I can add a comment.

It's up to you. Either you link the section in ISO, or you cna link this discussion here.

Reply:
> Either you link the section in ISO, or you cna link this discussion here.

Actually, in CPython docs it's documented here: https://docs.python.org/3.14/library/stdtypes.html#numeric-types-int-float-complex

"Python fully supports mixed arithmetic: when a binary arithmetic operator has operands of different numeric types, the operand with the “narrower” type is widened to that of the other, where integer is narrower than floating point, which is narrower than complex." (c)

Reply:
Actually, I think people would be more interested in understanding why the line `+= 0.0` is not replaced. `ci_result.sum` is a double, so they would say "but both of them are double, so adding should do nothing", but AFAIU, it's to transform `ci_result.sum == -0.0` to `ci_result.sum == 0.0` right?

Reply:
> Actually, I think people would be more interested in understanding why the line += 0.0 is needed at all.

Because you should add *both* components of promoted (to complex) float value.

> it's to transform ci_result.sum == -0.0 to ci_result.sum == 0.0 right?

Yes, this will affect just that corner case.

Reply:
If you want to add a comment, I think you can link the comment (or write your own explanation). Thank you again Sergey for the explanation!